lib/apiservers/engine/backends/container.go

```diff
@@ -582,9 +589,22 @@ func (c *Container) mapPorts(op portmap.Operation, hostconfig *container.HostCon
 
 		for _, pb := range pbs {
 			var hostPort int
-			hostPort, err = strconv.Atoi(pb.HostPort)
-			if err != nil {
-				return err
+			if pb.HostPort == "" {
+				// use a random port since no host port is specified
+				hostPort, err = requestHostPort(proto)
+				if err != nil {
+					log.Errorf("could not find available port on host")
+					return err
+				}
+				// update the hostconfig
+				hostpb := nat.PortBinding{HostIP: "", HostPort: strconv.Itoa(hostPort)}
+				hostconfig.PortBindings[nport] = []nat.PortBinding{hostpb}
```

Comment:
I think just setting `pb.HostPort` here would suffice; what you have here is replacing all the bindings.